File: Java/src/Modelo/algoritmos/EnhancedParallelBlock_IV.java
package Modelo.algoritmos;

public class EnhancedParallelBlock_IV implements Runnable{

    private static final int BLOCK_SIZE = 128;

    private double[][] A, B, C;
    private int size;

    public EnhancedParallelBlock_IV(double[][] A, double[][] B, double[][] C, int size)
    {
        this.A = A;
        this.B = B;
        this.C = C;
        this.size = size;
    }

    public void run() {
        for (int i1 = 0; i1 < size / 2; i1 += BLOCK_SIZE) {
            for (int j1 = 0; j1 < size; j1 += BLOCK_SIZE) {
                for (int k1 = 0; k1 < size; k1 += BLOCK_SIZE) {
                    for (int i = i1; i < i1 + BLOCK_SIZE && i < size; i++) {
                        for (int j = j1; j < j1 + BLOCK_SIZE && j < size; j++) {
                            for (int k = k1; k < k1 + BLOCK_SIZE && k < size; k++) {
                                A[i][k] += B[i][j] * C[j][k]; 
                            }
                        }
                    }
                }
            }
        }

        for (int i1 = size / 2; i1 < size; i1 += BLOCK_SIZE) {
            for (int j1 = 0; j1 < size; j1 += BLOCK_SIZE) {
                for (int k1 = 0; k1 < size; k1 += BLOCK_SIZE) {
                    for (int i = i1; i < i1 + BLOCK_SIZE && i < size; i++) {
                        for (int j = j1; j < j1 + BLOCK_SIZE && j < size; j++) {
                            for (int k = k1; k < k1 + BLOCK_SIZE && k < size; k++) {
                                A[i][k] += B[i][j] * C[j][k];
                            }
                        }
                    }
                }
            }
        }
    }
}

File: ProyectoFinal(Python)/modelo/algortimos/EnhancedParallelBlock_IV.py

class EnhancedParallelBlock_IV:

    BLOCK_SIZE = 128  # Define block size

    def run(self, A, B, C, size):
        for i1 in range(0, size // 2, self.BLOCK_SIZE):
            for j1 in range(0, size, self.BLOCK_SIZE):
                for k1 in range(0, size, self.BLOCK_SIZE):
                    for i in range(i1, min(i1 + self.BLOCK_SIZE, size)):
                        for j in range(j1, min(j1 + self.BLOCK_SIZE, size)):
                            for k in range(k1, min(k1 + self.BLOCK_SIZE, size)):
                                A[i][j] += B[i][k] * C[k][j]

        # Second half of the matrix (optional for parallelization with additional threads)
        for i1 in range(size // 2, size, self.BLOCK_SIZE):
            for j1 in range(0, size, self.BLOCK_SIZE):
                for k1 in range(0, size, self.BLOCK_SIZE):
                    for i in range(i1, min(i1 + self.BLOCK_SIZE, size)):
                        for j in range(j1, min(j1 + self.BLOCK_SIZE, size)):
                            for k in range(k1, min(k1 + self.BLOCK_SIZE, size)):
                                A[i][j] += B[i][k] * C[k][j]